harness/tests/determined/common/experimental/test_experiment.py

```diff
@@ -88,3 +89,30 @@ def test_wait_raises_exception_when_experiment_is_paused(
 
     with pytest.raises(ValueError):
         expref.wait()
+
+
+@responses.activate
+def test_list_trials_iter_paginated_responses(
```

Comment:
Thinking about what's being tested / what should be tested. Current test tests all of

1. limit is implemented properly
2. iterator doesn't stop prematurely
3. ensuing calls to the REST API only happen when items from previous pages are fully consumed

That seems good. And now that I'm reading back through the implementation you've got assertions for each of them. But I think they're also separate tests. 2, for example, is computed kind of implicitly from the assertion that's outside of the for loop (does that assertion check anything else?).

I've been pushing people lately across the team to write unit tests that test just one behavior, and to name those tests specifically after the expected behavior (like this one, we have heretofore most commonly named our tests after the feature under test rather than the behavior). Here, maybe `test_list_trials_iterates_through_all_trials`, `test_list_trials_requests_pages_lazily`, and `test_list_trials_respects_page_limit` (though could be that third test is unnecessary). If it's hard to come up with a name speaking to the behavior, that's a sign that the test is doing too much.

I think this is best practices. It makes it easy for a reader to understand at a glance what the tests cover (assuming they're implemented correctly), and when a test fails it's easy to understand why.

It _does_ have the downside of producing more code -- even when common functions are factored out. And it can also take more work to re-use expensive computation.

I think the upsides are worth the cost. And I _think_ the internet generally agrees. But there's room for differing opinions, and for sure this is just me asking for us to change.

So... I don't know. It's taken a lot of reinforcement, and I don't really know what other people think. Do you think we should talk about the practice of 1 behavior per test and test names that describe expected behavior?

Reply:
split these into two tests, `test_list_trials_respects_page_limit` doesn't seem like it belongs within this scope.